agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "php_agent.h"
+#include "php_user_instrument_hashmap.h"
+
+#if ZEND_MODULE_API_NO >= ZEND_7_4_X_API_NO
+
+typedef struct _nr_wraprecs_bucket {
+  struct _nr_wraprecs_bucket* prev;
+  struct _nr_wraprecs_bucket* next;
+  /* for efficiency wraprec is used as both: key and a value */
+  nruserfn_t* wraprec;
+} nr_wraprecs_bucket_t;
+
+typedef struct _nr_php_wraprec_hashmap {
+  nr_php_wraprec_hashmap_dtor_fn_t dtor_func;
+  size_t log2_num_buckets;
+  nr_wraprecs_bucket_t** buckets;
+  size_t elements;
+} nr_php_wraprec_hashmap_t;
+
+static inline size_t nr_count_buckets(const nr_php_wraprec_hashmap_t* hashmap) {
+  return (size_t)(1 << hashmap->log2_num_buckets);
+}
+
+static nr_php_wraprec_hashmap_t* nr_php_wraprec_hashmap_create_internal(
+    size_t log2_num_buckets,
+    nr_php_wraprec_hashmap_dtor_fn_t dtor_fn) {
+  nr_php_wraprec_hashmap_t* hashmap;
+
+  if (0 == log2_num_buckets) {
+    /*
+     * Encode the default value in one place: namely, here.
+     */
+    log2_num_buckets = 8;
+  } else if (log2_num_buckets > 24) {
+    /*
+     * Basic sanity check: it's extremely unlikely that we'll ever need a
+     * hashmap for the user function wraprecs that has more than 2^24 buckets.
+     */
+    log2_num_buckets = 24;
+  }
+
+  hashmap
+      = (nr_php_wraprec_hashmap_t*)nr_malloc(sizeof(nr_php_wraprec_hashmap_t));
+  hashmap->dtor_func = dtor_fn;
```

Comment:
Would `NULL` be a valid value for `dtor_func`? Is there any kind of sanity checking that should be done on this arg?

Reply:
`NULL` is a valid value for `dtor_func`. `dtor_func == NULL` means that whatever is stored in the hashmap does not have to be destroyed whenever hashmap bucket is destroyed. The sanity check is done [here](https://github.com/misiekn/newrelic-php-agent/blob/refactor/wraprec-lookup/agent/php_user_instrument_hashmap.c#L85)